libs/vkd3d/swapchain.c

```diff
@@ -1727,27 +1778,53 @@ static VkResult d3d12_swapchain_queue_present(struct d3d12_swapchain *swapchain,
 
     submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
     submit_info.pNext = NULL;
-    submit_info.waitSemaphoreCount = 0;
-    submit_info.pWaitSemaphores = NULL;
-    submit_info.pWaitDstStageMask = NULL;
+    submit_info.waitSemaphoreCount = 1;
+    assert(swapchain->vk_acquire_semaphores_signaled[swapchain->frame_id]);
+    submit_info.pWaitSemaphores = &swapchain->vk_acquire_semaphores[swapchain->frame_id];
+    submit_info.pWaitDstStageMask = &acquire_wait_mask;
     submit_info.commandBufferCount = 1;
     submit_info.pCommandBuffers = &vk_cmd_buffer;
     submit_info.signalSemaphoreCount = 1;
-    submit_info.pSignalSemaphores = &swapchain->vk_semaphores[swapchain->vk_image_index];
+    submit_info.pSignalSemaphores = &swapchain->vk_present_semaphores[swapchain->vk_image_index];
 
     if ((vr = vk_procs->vkQueueSubmit(vk_queue, 1, &submit_info, VK_NULL_HANDLE)) < 0)
     {
         ERR("Failed to blit swapchain buffer, vr %d.\n", vr);
         return vr;
     }
 
+    swapchain->vk_acquire_semaphores_signaled[swapchain->frame_id] = false;
+
     present_info.waitSemaphoreCount = 1;
-    present_info.pWaitSemaphores = &swapchain->vk_semaphores[swapchain->vk_image_index];
+    present_info.pWaitSemaphores = &swapchain->vk_present_semaphores[swapchain->vk_image_index];
 
     if ((vr = vk_procs->vkQueuePresentKHR(vk_queue, &present_info)) >= 0)
     {
+        swapchain->frame_id = (swapchain->frame_id + 1) % swapchain->buffer_count;
         swapchain->vk_image_index = INVALID_VK_IMAGE_INDEX;
-        /* Could get SUBOPTIMAL here. Deal with it later. */
+
+        /* Could get SUBOPTIMAL here. Defer acquiring if we hit that path.
+         * On next present, we can recreate the swapchain. */
+        if (vr == VK_SUCCESS)
+        {
+            /* Try to acquire our next image to avoid blocking next frame. */
+            assert(!swapchain->vk_acquire_semaphores_signaled[swapchain->frame_id]);
+            vr = vk_procs->vkAcquireNextImageKHR(vk_device, swapchain->vk_swapchain, UINT64_MAX,
+                    swapchain->vk_acquire_semaphores[swapchain->frame_id], VK_NULL_HANDLE,
+                    &swapchain->vk_image_index);
+
+            if (vr >= 0)
+            {
+                swapchain->vk_acquire_semaphores_signaled[swapchain->frame_id] = true;
```

Comment:
We still want to `swapchain->vk_image_index = INVALID_VK_IMAGE_INDEX;` here

Reply:
Do we though? The next present will call Acquire again, while we have an image acquired.